### ISSUE TITLE
Possible bug in Dijkstra's code

### ISSUE THREAD

**Issue by rcasero** [rcasero/gerardus#89]
Reported by direct email by Matthew Morena: 

"Attached are several matlab matrices (see dijkstra_algebra_mwe_morena.mat), one of which is the weighted adjacency matrix, WeightAdjMatrix, that I have been using to compare our algorithms.  This is a 4002x4002 integer-valued matrix.  I use this matrix to generate a DistanceMatrix and a PredecessorMatrix using a slow (e.g., non-Fibonacci heap) version of Dijkstra's algorithm (see the attached Dijkstra_Algorithm.m and matlab diary), and I use this matrix to generate corresponding distance and predecessor matrices using the implementation of Dijkstra's algorithm from the Gerardus collection.  I am also attaching a matlab diary file indicating which commands I use to compare our results.  I should point out that I am using version -r248 of the Gerardus dijkstra algorithm, having checked out the files via:

svn co -r248 http://gerardus.googlecode.com/svn/trunk/matlab/ThirdPartyToolbox/ dijkstra_fibonacci_heap

I checked out -r248 for two reasons: 
I could not get the latest version, -r1712, to work on my MacBook Pro 10.8 because of issues first with installing and compiling (via cake) InsightToolkit-4.8.0, and then with mex-ing dijkstra.cpp after copying all of the relevant ITK*.h files.
I have no need and am not familiar with the "piggyback extension" in my research, dijkstra.cpp from version -r1712.

I hope that the matlab diary is clear enough to show you that the two distance matrices are essentially equivalent (one is simply the transpose of the other), but that significant differences can be found between the two predecessor matrices.  I have stepped through the code in dijkstra.cpp, but I am not familiar enough with the fibonacci heap to solve these discrepancies.  I will say that the fibonacci heap is incredibly fast and efficient, making me very curious to see if I can get the results of the different Dijkstra's routines to match up.

To summarize: my matlab code does not return the same predecessor matrix as your cpp code; however, the two distance matrices are equivalent.  If you have time and feel comfortable taking a look at this situation, I would very much appreciate it!"
[21-Jul-2015 copy.txt](https://github.com/rcasero/gerardus/files/2585/21-Jul-2015.copy.txt)

**Comment by mamorena** [comment]
Thank you for posting my questions.  Any help by any GitHub users would be much appreciated!

**Comment by rcasero** [comment]
First test: From our [test script for dijkstra.m](https://github.com/rcasero/gerardus/blob/master/matlab/test/test_dijkstra.m#L38), we have symmetric distance matrix

d = [...
    %  1   2   3   4   5   6
       0   7   9   0   0  14; ...
       7   0  10  15   0   0; ...
       9  10   0  11   0   2; ...
       0  15  11   0   6   0; ...
       0   0   0   6   0   9; ...
      14   0   2   0   9   0];

Running the gerardus dijkstra.mexa64 and Matt's Dijkstra_Algorithm() for each node as the source produces the same result, in terms of  distance path and predecessors. So no discrepancy there. Let's see other examples.

**Comment by rcasero** [comment]
OK, the problem happened only with matrices that were not symmetric. The MEX C++ implementation we were using assumes that G(j,i) is the distance from i to j, but we were using the transpose.

The solution is simply to transpose the input sparse matrix. I have updated the help header in dijkstra.m to note that now G(j,i) means the distance from node i to node j.

I have also recoded the MEX function dijkstra.cpp so that the outputs are given with the same format (i.e. by columns, rather than rows).

The fix for this bug is
 Commit: 6d4b74e8999d6b38697d5373249a21a19fa4f060
      https://github.com/rcasero/gerardus/commit/6d4b74e8999d6b38697d5373249a21a19fa4f060
